### PR TITLE
Fix memory issues in ImageIO.c

### PR DESCRIPTION
**Pull request by JonathanMeans**

### CODE REVIEW

File: modules/c/nitf/source/ImageIO.c [hunk 1]
```diff
@@ -5855,12 +5855,16 @@ NITFPRIV(void) nitf_ImageIOControl_destruct(_nitf_ImageIOControl ** cntl)
...
         nBlockCols = cntlActual->nBlockIO / cntlActual->numBandSubset;
-        blocks = &(cntlActual->blockIO[0][0]);
+
         for (i = 0; i < nBlockCols; ++i)
         {
-            if (blocks[i].blockControl.freeFlag)
+            for (j = 0; j < cntlActual->numBandSubset; ++j)
             {
-                NITF_FREE(blocks[i].blockControl.block);
+                blocks = &(cntlActual->blockIO[i][j]);
```
Comment (asylvest) [inline modules/c/nitf/source/ImageIO.c]: So basically this previously leaked for multi-band NITFs?

File: modules/c/nitf/source/ImageIO.c [hunk 2]
```diff
@@ -7165,6 +7169,7 @@ NITFPRIV(int) nitf_ImageIO_writeToBlock(_nitf_ImageIOBlock * blockIO,
                              NITF_STRERROR(NITF_ERRNO));
             return NITF_FAILURE;
         }
+        memset(blockCntl->block, 0, nitf->blockSize);
```
Comment (asylvest) [inline modules/c/nitf/source/ImageIO.c]: This is just a big scratch buffer right?  So I don't think we need to actually `memset` it upfront...

File: modules/c/nitf/source/ImageIO.c [hunk 3]
```diff
@@ -9482,14 +9480,16 @@ nitf_ImageIO_12PixelComStart(nitf_CompressionControl *object,
...
   icntl->padMask = padMask;
-  icntl->buffer = NULL;
   icntl->written = 0;
 
 /* Allocate compressed block buffer */
 
-  icntl->buffer = (nitf_Uint8 *) NITF_MALLOC(icntl->blockSizeCompressed);
-  if(icntl->buffer == NULL)
-    return(NITF_FAILURE);
+  if (!icntl->buffer)
```
Comment (asylvest) [inline modules/c/nitf/source/ImageIO.c]: Same difference but change this check to `icntl->buffer == NULL` just for code legibility

File: modules/c/nitf/source/ImageIO.c [hunk 4]
```diff
@@ -9482,14 +9480,16 @@ nitf_ImageIO_12PixelComStart(nitf_CompressionControl *object,
...
+  {
+    icntl->buffer = (nitf_Uint8 *) NITF_MALLOC(icntl->blockSizeCompressed);
+    if(icntl->buffer == NULL)
```
Comment (asylvest) [inline modules/c/nitf/source/ImageIO.c]: Need a space between the `if` and the first paren

File: modules/c/nitf/source/ImageIO.c (hunk 2 above)
Comment (JonathanMeans) [reply modules/c/nitf/source/ImageIO.c]: That's what I assumed, but valgrind complains a lot without it.

    ==29705== Conditional jump or move depends on uninitialised value(s)
    ==29705==    at 0xCBABE71: opj_t1_encode_cblk (t1.c:2288)
    ==29705==    by 0xCBABC79: opj_t1_encode_cblks (t1.c:2192)
    ==29705==    by 0xCB8338D: opj_tcd_t1_encode (tcd.c:2505)
    ==29705==    by 0xCB80DDE: opj_tcd_encode_tile (tcd.c:1408)
    ==29705==    by 0xCB5E4E9: opj_j2k_write_sod (j2k.c:4680)
    ==29705==    by 0xCB6F459: opj_j2k_write_first_tile_part (j2k.c:11754)
    ==29705==    by 0xCB6ECCD: opj_j2k_post_write_tile (j2k.c:11513)
    ==29705==    by 0xCB6FEA4: opj_j2k_write_tile (j2k.c:12108)
    ==29705==    by 0xCB574EE: opj_write_tile (openjpeg.c:916)
    ==29705==    by 0xCB3EA5D: OpenJPEGWriter_setTile (OpenJPEGImpl.c:1028)
    ==29705==    by 0xCB3F91B: j2k_Writer_setTile (Writer.c:57)
    ==29705==    by 0xCB3C49D: implWriteBlock (J2KCompress.c:376)

File: modules/c/nitf/source/ImageIO.c (hunk 1 above)
Comment (JonathanMeans) [reply modules/c/nitf/source/ImageIO.c]: Yes

File: modules/c/nitf/source/ImageIO.c (hunk 2 above)
Comment (asylvest) [reply modules/c/nitf/source/ImageIO.c]: Hmm... that indicates to me that we're not actually filling in some image content (or something else) fully before calling the compression.  Can you try to track down if the test program isn't fully filling in the input image or where this is coming from (may be easiest to set the buffer to 255 everywhere then try to find where/when/why you still see 255s by the time you get to compression)?

Comment (asylvest) [reply modules/c/nitf/source/ImageIO.c]: Punting on this one for now - will address in future PR

**Pull request merged**